pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/rebalance/TableRebalancer.java

```diff
@@ -0,0 +1,405 @@
+    if (tableConfig.getTableType() == TableType.OFFLINE) {
+      instancePartitionsMap.put(InstancePartitionsType.OFFLINE,
+          getInstancePartitions(tableConfig, InstancePartitionsType.OFFLINE, reassignInstances, dryRun));
+    } else {
+      instancePartitionsMap.put(InstancePartitionsType.CONSUMING,
+          getInstancePartitions(tableConfig, InstancePartitionsType.CONSUMING, reassignInstances, dryRun));
+      instancePartitionsMap.put(InstancePartitionsType.COMPLETED,
+          getInstancePartitions(tableConfig, InstancePartitionsType.COMPLETED, reassignInstances, dryRun));
+    }
+
+    LOGGER.info("Calculating the target assignment for table: {}", tableNameWithType);
+    SegmentAssignment segmentAssignment = SegmentAssignmentFactory.getSegmentAssignment(_helixManager, tableConfig);
+    Map<String, Map<String, String>> currentAssignment = currentIdealState.getRecord().getMapFields();
+    Map<String, Map<String, String>> targetAssignment =
+        segmentAssignment.rebalanceTable(currentAssignment, instancePartitionsMap, rebalanceConfig);
+
+    if (currentAssignment.equals(targetAssignment)) {
+      LOGGER.info("Table: {} is already balanced", tableNameWithType);
+      if (reassignInstances) {
+        return new RebalanceResult(RebalanceResult.Status.DONE, "Instance reassigned, table is already balanced",
+            instancePartitionsMap, targetAssignment);
+      } else {
+        return new RebalanceResult(RebalanceResult.Status.NO_OP, "Table is already balanced", instancePartitionsMap,
+            targetAssignment);
+      }
+    }
+
+    if (dryRun) {
+      LOGGER.info("Rebalance table: {} in dry-run mode, returning the target assignment", tableNameWithType);
+      return new RebalanceResult(RebalanceResult.Status.DONE, "Dry-run mode", instancePartitionsMap, targetAssignment);
+    }
+
+    int minAvailableReplicas;
+    if (rebalanceConfig.getBoolean(RebalanceConfigConstants.DOWNTIME, RebalanceConfigConstants.DEFAULT_DOWNTIME)) {
+      minAvailableReplicas = 0;
+      LOGGER.info("Rebalancing table: {} with downtime", tableNameWithType);
+    } else {
+      minAvailableReplicas = rebalanceConfig.getInt(RebalanceConfigConstants.MIN_REPLICAS_TO_KEEP_UP_FOR_NO_DOWNTIME,
+          RebalanceConfigConstants.DEFAULT_MIN_REPLICAS_TO_KEEP_UP_FOR_NO_DOWNTIME);
+      int numCurrentReplicas = currentAssignment.values().iterator().next().size();
+      int numTargetReplicas = targetAssignment.values().iterator().next().size();
+      // Use the smaller one to determine the min available replicas
+      int numReplicas = Math.min(numCurrentReplicas, numTargetReplicas);
+      if (minAvailableReplicas > 0) {
+        if (minAvailableReplicas >= numReplicas) {
+          LOGGER.warn(
+              "Illegal config for min available replicas: {} for table: {}, must be less than number of replicas (current: {}, target: {})",
+              minAvailableReplicas, tableNameWithType, numCurrentReplicas, numTargetReplicas);
```

Comment:
In this case can we just implicitly default to min available replicas as 1, stil log the message though but continue with rebalancing?

Reply:
That might cause performance issue, so I would rather be conservative and fail the rebalance. User can fix there config and rerun the rebalance